File: src/core/filelog.ts
import { Platform } from 'react-native';
import * as FileSystem from 'expo-file-system';

/**
 * Persistent, FILE-BASED JS diagnostic logger — the JS-side companion to the
 * native `FileLog` (modules/killio-speech/.../FileLog.kt). It captures the JS
 * half of the wake/diary pipeline (native-wake handling, keyword push, capture
 * start/stop, diary enqueue) into a file that survives app restarts and can be
 * pulled/shared WITHOUT adb/logcat.
 *
 * Where it writes:
 *   We want the SAME external folder the native logger uses
 *   (/sdcard/Android/data/<pkg>/files/logs/), but expo-file-system's sandboxed
 *   `documentDirectory` does NOT point at the app's external files dir, and there
 *   is no public expo API to resolve getExternalFilesDir() from JS. So the JS log
 *   is written to:
 *
 *     <FileSystem.documentDirectory>logs/killio-js.log
 *
 *   On Android `documentDirectory` is the app's INTERNAL files dir
 *   (/data/data/dev.killio.vault/files/), which is NOT world-pullable without
 *   root. Two ways to get it off-device:
 *     1. In-app "Exportar logs" button (app/settings.tsx) shares this file via
 *        expo-sharing — works with NO adb. (Primary path.)
 *     2. adb (debuggable build):  adb exec-out run-as dev.killio.vault \
 *          cat files/logs/killio-js.log
 *
 *   The native KWS/capture log (the priority for debugging the wake) DOES land in
 *   the adb-pullable external dir; this JS log complements it.
 *
 * Best-effort everywhere: every call swallows errors and never throws, so a
 * logging failure can never break capture or the app. Writes are serialized
 * through a promise chain (RN JS is single-threaded but async appends could
 * interleave) and the file is size-capped with a simple 2-file rotation.
 */

const LOG_DIR = `${FileSystem.documentDirectory ?? ''}logs/`;
const LOG_FILE = `${LOG_DIR}killio-js.log`;
const ROTATED_FILE = `${LOG_DIR}killio-js.1.log`;
const MAX_BYTES = 2 * 1024 * 1024; // ~2MB cap before rotation

/** Absolute on-disk path of the JS log file (file:// URI). */
export const JS_LOG_PATH = LOG_FILE;
export const JS_LOG_ROTATED_PATH = ROTATED_FILE;
export const JS_LOG_DIR = LOG_DIR;

let dirReady = false;
/** Serialize appends so concurrent flog() calls don't clobber each other. */
let writeChain: Promise<void> = Promise.resolve();
let headerWritten = false;

async function ensureDir(): Promise<boolean> {
  if (!FileSystem.documentDirectory) return false;
  if (dirReady) return true;
  try {
    const info = await FileSystem.getInfoAsync(LOG_DIR);
    if (!info.exists) {
      await FileSystem.makeDirectoryAsync(LOG_DIR, { intermediates: true });
    }
    dirReady = true;
    return true;
  } catch {
    return false;
  }
}

/** Rename to killio-js.1.log (keep 2 files) once over the size cap. */
async function rotateIfNeeded(): Promise<void> {
  try {
    const info = await FileSystem.getInfoAsync(LOG_FILE);
    if (!info.exists || (info.size ?? 0) < MAX_BYTES) return;
    await FileSystem.deleteAsync(ROTATED_FILE, { idempotent: true });
    // expo-file-system has no rename; move via copy+delete.
    await FileSystem.copyAsync({ from: LOG_FILE, to: ROTATED_FILE });
    await FileSystem.deleteAsync(LOG_FILE, { idempotent: true });
  } catch {
    /* ignore — a failed rotation must not break logging */
  }
}

function isoNow(): string {
  return new Date().toISOString(); // already ISO8601 with millis + 'Z'
}

async function appendLine(line: string): Promise<void> {
  if (!(await ensureDir())) return;
  await rotateIfNeeded();
  // expo-file-system has no append mode; read-modify-write. At our log rate
  // (a handful of lines per wake/capture event) this is fine.
  let prev = '';
  try {
    const info = await FileSystem.getInfoAsync(LOG_FILE);
    if (info.exists) {
      prev = await FileSystem.readAsStringAsync(LOG_FILE, {
        encoding: FileSystem.EncodingType.UTF8,
      });
    }
  } catch {
    prev = '';
  }
  await FileSystem.writeAsStringAsync(LOG_FILE, prev + line, {
    encoding: FileSystem.EncodingType.UTF8,
  });
}

/**
 * Append a timestamped line `<ISO8601Z> <TAG>: <message>` to killio-js.log.
 * Best-effort + fire-and-forget — never throws, never blocks the caller. On a
 * platform without a document dir (web) it silently no-ops.
 */
export function flog(tag: string, msg: string): void {
  // Mirror to the JS console too so a live `npx react-native log-android` still
  // shows it (tee — same spirit as the native Log.i + FileLog tee).
  try {
    console.log(`[flog:${tag}] ${msg}`);
  } catch {
    /* ignore */
  }
  if (!FileSystem.documentDirectory) return;
  const line = `${isoNow()} ${tag}: ${msg}\n`;
  writeChain = writeChain
    .then(async () => {
      if (!headerWritten) {
        headerWritten = true;
        await appendLine(
          `${isoNow()} filelog: ──────── JS flog init — js log at ${LOG_FILE} (platform=${Platform.OS}) ────────\n`,
        );
      }
      await appendLine(line);
    })
    .catch(() => {
      /* swallow — logging must never crash the caller */
    });
}

/**
 * Read the full JS log (current + rotated, oldest first) as a single string for
 * the in-app export/share button. Returns '' if nothing is logged yet.
 */
export async function readJsLog(): Promise<string> {
  const parts: string[] = [];
  for (const p of [ROTATED_FILE, LOG_FILE]) {
    try {
      const info = await FileSystem.getInfoAsync(p);
      if (info.exists) {
        parts.push(
          await FileSystem.readAsStringAsync(p, {
            encoding: FileSystem.EncodingType.UTF8,
          }),
        );
      }
    } catch {
      /* ignore unreadable part */
    }
  }
  return parts.join('');
}

File: modules/killio-speech/android/src/main/java/dev/killio/speech/FileLog.kt
package dev.killio.speech

import android.content.Context
import android.util.Log
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Persistent, FILE-BASED diagnostic logger that survives app restarts and needs
 * NO adb/logcat to read. Logs are written to the app's EXTERNAL files dir:
 *
 *   /sdcard/Android/data/<pkg>/files/logs/killio.log
 *
 * which is pullable WITHOUT root via:
 *   adb pull /sdcard/Android/data/dev.killio.vault/files/logs/
 * and also survives across app restarts. The JS side (src/core/filelog.ts)
 * writes its own killio-js.log into the SAME folder, and the in-app
 * "Exportar logs" button (app/settings.tsx) shares both files.
 *
 * Design goals (per the logging spec):
 *  - BEST-EFFORT: a logging failure must NEVER throw or crash capture. Every
 *    public method swallows all exceptions.
 *  - Thread-safe: all writes are serialized on a lock (the recognition loop, the
 *    KWS feed and the one-shot path may all log concurrently).
 *  - Rolling: one file `killio.log`, capped at ~2MB; when it exceeds the cap it
 *    is renamed to `killio.1.log` (the previous .1 is overwritten), so at most
 *    two files (~4MB) are kept.
 *  - Line format: `<ISO8601Z> <TAG>: <message>` e.g.
 *      2026-06-10T14:03:11.482Z KillioKWS: WAKE keyword="hey killio"
 *
 * Usage: call [init] once (VaultSpeechService.onCreate / onStartCommand) to
 * resolve the external dir, then [log] from anywhere. The native KWS/STT
 * diagnostics in VaultSpeechService are TEE'd here (Log.i kept + FileLog.log
 * added) so the wake/capture pipeline can be debugged by reading the file.
 */
object FileLog {
  private const val LOG_NAME = "killio.log"
  private const val ROTATED_NAME = "killio.1.log"
  private const val MAX_BYTES = 2L * 1024 * 1024 // ~2MB cap before rotation

  private val lock = Any()

  @Volatile private var logsDir: File? = null
  @Volatile private var logFile: File? = null

  /** ISO8601 with millis in UTC (Z). Guarded by [lock] (SimpleDateFormat is not
   *  thread-safe). */
  private val iso = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
    timeZone = TimeZone.getTimeZone("UTC")
  }

  /**
   * Resolve the external logs dir once and write a header line. Idempotent and
   * best-effort: safe to call on every onStartCommand. Falls back to the
   * internal filesDir/logs if the external dir is unavailable (e.g. no SD
   * mounted) so logging never silently no-ops.
   */
  fun init(context: Context) {
    if (logFile != null) return
    synchronized(lock) {
      if (logFile != null) return
      try {
        // Prefer the EXTERNAL files dir (adb-pullable without root). Fall back to
        // the internal filesDir if external storage isn't available.
        val base = context.getExternalFilesDir(null) ?: context.filesDir
        val dir = File(base, "logs").apply { mkdirs() }
        logsDir = dir
        val f = File(dir, LOG_NAME)
        logFile = f
        appendRaw(
          f,
          line(
            "FileLog",
            "──────── FileLog init — native logs at ${f.absolutePath} (pull: adb pull ${dir.absolutePath}) ────────",
          ),
        )
        // Mirror the resolved path to logcat once so it's discoverable there too.
        Log.i("FileLog", "native log file: ${f.absolutePath}")
      } catch (_: Throwable) {
        // Never throw from logging init.
      }
    }
  }

  /** Absolute path of the on-disk log file (or null if not yet initialized). */
  fun currentPath(): String? = logFile?.absolutePath

  /** Absolute path of the logs directory (or null if not yet initialized). */
  fun currentDir(): String? = logsDir?.absolutePath

  /**
   * Return the TAIL of the native log (current + rotated, oldest-first), capped at
   * ~[maxBytes] so the in-app "Export logs" can concatenate the native lines (KWS +
   * whisper download + capture) with the JS log WITHOUT adb. Resolves the external
   * file even if [init] hasn't been called yet (e.g. capture never started this
   * launch) by best-effort re-resolving from [context]. Never throws; returns "" if
   * nothing is readable.
   */
  fun readTail(context: Context, maxBytes: Int = 200 * 1024): String {
    return try {
      // Resolve the dir even if the speech service never ran this launch.
      val dir = logsDir ?: run {
        val base = context.getExternalFilesDir(null) ?: context.filesDir
        File(base, "logs")
      }
      val current = logFile ?: File(dir, LOG_NAME)
      val rotated = File(dir, ROTATED_NAME)
      val sb = StringBuilder()
      synchronized(lock) {
        // oldest first: rotated then current.
        for (f in listOf(rotated, current)) {
          if (f.exists() && f.length() > 0) {
            try { sb.append(f.readText()) } catch (_: Throwable) {}
          }
        }
      }
      val text = sb.toString()
      if (text.length <= maxBytes) text
      else "…(truncated)…\n" + text.substring(text.length - maxBytes)
    } catch (_: Throwable) {
      ""
    }
  }

  /**
   * Append a timestamped line `<ISO8601Z> <TAG>: <message>` to killio.log.
   * Thread-safe, buffered (one open/append per call — fine at our log rate),
   * best-effort — never throws. No-op if [init] hasn't resolved a file yet.
   */
  fun log(tag: String, msg: String) {
    val f = logFile ?: return
    try {
      synchronized(lock) {
        rotateIfNeeded(f)
        appendRaw(f, line(tag, msg))
      }
    } catch (_: Throwable) {
      // swallow — logging must never crash the caller.
    }
  }

  // ── internals (all called under [lock]) ──────────────────────────────────

  private fun line(tag: String, msg: String): String =
    "${iso.format(Date())} $tag: $msg\n"

  private fun appendRaw(f: File, text: String) {
    // append=true; UTF-8 default. Open/close per write keeps it simple and means
    // a crash can never leave a half-buffered file.
    f.appendText(text)
  }

  /** Rename to killio.1.log (keep 2 files) once over the size cap. */
  private fun rotateIfNeeded(f: File) {
    try {
      if (f.length() < MAX_BYTES) return
      val rotated = File(f.parentFile, ROTATED_NAME)
      if (rotated.exists()) rotated.delete()
      if (!f.renameTo(rotated)) {
        // rename failed (rare) — truncate in place rather than grow unbounded.
        f.delete()
      }
    } catch (_: Throwable) {
      // ignore — a failed rotation must not break logging.
    }
  }
}
